#!/usr/bin/env Rscript

library(rtracklayer)
library(GenomicRanges)
library(gUtils)
library(BSgenome.Mmusculus.UCSC.mm10)
library(stringr)
library(parallel)

## bypass X11 rendering 
options(bitmapType='cairo')

args = commandArgs(trailingOnly=TRUE)

if (length(args) != 7) {
    stop("Incorrect number of arguments!\nUsage:\n> augment_transcriptome.R <cores> <upstreamUTRFile> <downstreamUTRFile> <annotFile> <extendDownstream> <resultsPrefix> <resultsSuffix>\n")
}

arg.cores <- as.integer(args[1])
arg.upstreamUTRFile <- args[2]
arg.downstreamUTRFile <- args[3]
arg.annotFile <- args[4]
arg.extDownstream <- as.integer(args[5])
arg.outPrefix <- args[6]
arg.outSuffix <- args[7]


## Load References
mm10 <- BSgenome.Mmusculus.UCSC.mm10

## Load Files
cat("Loading input files...\n")
annotation.gr <- keepStandardChromosomes(import(arg.annotFile, genome='mm10'))
upstreamUTR.sites.gr <- keepStandardChromosomes(import(arg.upstreamUTRFile, genome='mm10'))
extendedUTR.sites.gr <- keepStandardChromosomes(import(arg.downstreamUTRFile, genome='mm10'))


adjust.transcriptEnd <- function (tx.id, end.new, tx.strand) {
    stopifnot(tx.strand %in% c('-', '+'))

    tx.gr <- annotation.gr %Q% (!is.na(transcript_id) & transcript_id == tx.id)
    tx <- tx.gr[tx.gr$type == 'transcript',]

    ## LABELS
    offset <- if (tx.strand == '+') {
                  end.new - end(gr.end(tx, ignore.strand = FALSE))
              } else {
                  end(gr.end(tx, ignore.strand = FALSE)) - end.new
              }
    ## update transcript_id
    tx.id.new <- paste0(tx$transcript_id, "-UTR", if (offset < 0) "" else "+", offset)
    elementMetadata(tx.gr)[,"transcript_id"] <- tx.id.new
    ## update transcript_name
    tx.name.new <- paste0(tx$transcript_name, "-UTR", if (offset < 0) "" else "+", offset)
    elementMetadata(tx.gr)[,"transcript_name"] <- tx.name.new
    ## update ID
    elementMetadata(tx.gr)[, "ID"] <- str_replace(tx.gr$ID, tx.id, tx.id.new)
    ## update Parent
    elementMetadata(tx.gr)[, "Parent"] <- str_replace(tx.gr$Parent, tx.id, tx.id.new)

    ## ADJUST TX END
    end.cur <- end(gr.end(tx, ignore.strand = FALSE))
    ##cat(sprintf("Adjusting %s on %s strand from %d to %d\n", tx.id, tx.strand, end.cur, end.new))

    if (offset > 0) {
        if (tx.strand == '+') {
            ## Find all elements contiguous w/ TX end and extend range end
            end(tx.gr[end(gr.end(tx.gr, ignore.strand = FALSE)) == end.cur,]) <- end.new
        } else {
            ## Find all elements contiguous w/ TX end and extend range end
            start(tx.gr[end(gr.end(tx.gr, ignore.strand = FALSE)) == end.cur,]) <- end.new
        }
    } else {
        if (tx.strand == "+") {
            ## Delete all elements that might get removed (e.g., if multiple exons in 3' UTR)
            tx.gr <- tx.gr[end(gr.start(tx.gr, ignore.strand = FALSE)) < end.new,]
            ## For remaining ranges, clip at new site
            end(tx.gr[end(gr.end(tx.gr, ignore.strand = FALSE)) > end.new,]) <- end.new
        } else {
            ## Delete all elements that might get removed (e.g., if multiple exons in 3' UTR)
            tx.gr <- tx.gr[end(gr.start(tx.gr, ignore.strand = FALSE)) > end.new,]
            ## For remaining ranges, clip at new site
            start(tx.gr[end(gr.end(tx.gr, ignore.strand = FALSE)) < end.new,]) <- end.new
        }
    }

    tx.gr
}

## Process Upstream UTR Sites
cat("Intersecting annotation with upstream sites...\n")
upstream.txs.gr <- gr.findoverlaps(
    upstreamUTR.sites.gr,
    annotation.gr %Q% (type == 'three_prime_UTR' & transcript_type == 'protein_coding'),
    scol = c("transcript_id"), ignore.strand = FALSE)

if (all(1:length(upstreamUTR.sites.gr) %in% upstream.txs.gr$query.id)) {
    cat("All cleavage sites intersected known 3' UTRs\n")
} else {
    warning("Some cleavage sites in upstreamUTR.sites.gr did not intersect known sites!")
}

cat("Extracting transcripts for upstream sites...\n")
upstream.annotation.gr <- grbind(
    mcmapply(adjust.transcriptEnd,
             tx.id = upstream.txs.gr$transcript_id,
             end.new = end(gr.end(upstream.txs.gr, ignore.strand = FALSE)),
             tx.strand = as.character(strand(upstream.txs.gr)),
             mc.cores = arg.cores, SIMPLIFY = FALSE)
)

cat("Exporting transcriptome for upstream sites...\n")
upstream.annotation.gff3.file <- sprintf("data/gff/%s.txs.utr3.%s.gff3", arg.outPrefix, arg.outSuffix)
upstream.annotation.gtf.file <- sprintf("data/gff/%s.txs.utr3.%s.gtf", arg.outPrefix, arg.outSuffix)
export.gff3(upstream.annotation.gr, upstream.annotation.gff3.file)
export(upstream.annotation.gr, upstream.annotation.gtf.file, format = "gtf")

cat("Clearing cached files...\n")
rm(list=c("upstream.annotation.gr", "upstream.txs.gr", "upstreamUTR.sites.gr"))

    
## Process Extended UTR Sites
cat("Intersecting annotation with downstream sites...\n")
extended.txs.gr <- gr.findoverlaps(
    extendedUTR.sites.gr,
    flank(gr.end(annotation.gr %Q%
                 (type == 'three_prime_UTR' &
                  transcript_type == 'protein_coding'),
                 ignore.strand = FALSE),
          width = arg.extDownstream, start = FALSE),
    scol = c("transcript_id"), ignore.strand = FALSE)

if (all(1:length(extendedUTR.sites.gr) %in% extended.txs.gr$query.id)) {
    cat("All cleavage sites intersected within 5Kb downstream of known 3' UTRs\n")
} else {
    warning("Some cleavage sites in extendedUTR.sites.gr did not intersect known sites!")
}

cat("Extracting transcripts for extended transcriptome...\n")
extended.annotation.gr <- grbind(
    mcmapply(adjust.transcriptEnd,
             tx.id = extended.txs.gr$transcript_id,
             end.new = end(gr.end(extended.txs.gr, ignore.strand = FALSE)),
             tx.strand = as.character(strand(extended.txs.gr)),
             mc.cores = arg.cores, SIMPLIFY = FALSE)
)

cat("Exporting transcripts for downstream sites...\n")
extended.annotation.gff3.file <- sprintf("data/gff/%s.txs.extutr3.%s.gff3", arg.outPrefix, arg.outSuffix)
extended.annotation.gtf.file <- sprintf("data/gff/%s.txs.extutr3.%s.gtf", arg.outPrefix, arg.outSuffix)
export.gff3(extended.annotation.gr, extended.annotation.gff3.file)
export(extended.annotation.gr, extended.annotation.gtf.file, format = "gtf")

cat("All operations complete.")
