---
title: "Summarize PolyASite Cleavage Sites"
author: "Mervin Fansler"
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  html_document:
    toc: true
    toc_float: true
---

<!-- Mock Data -->
```{r mock_data, include=FALSE, eval=FALSE}
if (interactive()) {
    Snakemake <- setClass("Snakemake", slots=c(input='list', output='list', params='list'))
    snakemake <- Snakemake(
        input=list(pas="/data/mayrc/db/mm10/polyASite.clusters.mm10.bed.gz",
                   gencode="/data/mayrc/db/mm10/gencode.vM21.annotation.mRNA_ends_found.gff3.gz"),
        params=list(genome='mm10',
                    extUpstream="1000",
                    extDownstream="5000"))
}

## bypass X11 rendering 
options(bitmapType='cairo')
```

# Libraries
```{r libs, message=FALSE, warning=FALSE}
library(rtracklayer)
library(GenomicRanges)
library(GenomicFeatures)
library(gUtils)
library(BSgenome)
library(tidyverse)
library(magrittr)
```

# Load Data
```{r load_data}
genome <- getBSgenome(snakemake@params$genome)

gencode.gr <- keepStandardChromosomes(import(snakemake@input$gencode,
                                             genome=snakemake@params$genome),
                                      pruning.mode="coarse")

polyASites.df <- read_tsv(snakemake@input$pas, na=c("NA", "."),
                          col_names=c("chrom", "start", "stop", "name",
                                      "score", "strand", "pA_signal", "gene"),
                          col_types="ciiciccc") %>%
    mutate(annot=str_extract(name, '[^:]+$'))

polyASites.gr <- makeGRangesFromDataFrame(polyASites.df, keep.extra.columns=TRUE,
                                          seqinfo=seqinfo(genome)) %>%
    `names<-`(.$name)

```

# Plot Score Distribution
```{r fig.width=6, fig.height=4}
polyASites.df %>%
    ggplot(aes(x=as.character(score), fill=annot)) +
    geom_bar() +
    labs(x="Supporting Protocols",
         y="Cleavage Sites",
         fill="Annotation") +
    theme_minimal()
```


```{r fig.width=8, fig.height=8}
polyASites.df %>%
    ggplot(aes(x=as.character(score), fill=is.na(pA_signal))) +
    geom_bar(position='fill') +
    labs(x="Supporting Protocols",
         y="Cleavage Sites",
         fill="No PAS") +
    scale_fill_manual(values=c('FALSE'='lightgrey', 'TRUE'='red')) +
    facet_wrap(vars(annot)) +
    theme_minimal()
```

# Intersect GENCODE
```{r}
pc_txs.gr <- gencode.gr %Q% (type == 'transcript' &
                             transcript_type == 'protein_coding')
nc_txs.gr <- gencode.gr %Q% (type == 'transcript' &
                             transcript_type != 'protein_coding')

pc_pas.gr <- (polyASites.gr + 25) %&&%
    gr.end(pc_txs.gr, ignore.strand=FALSE, width=25)

npc_pas.gr <- polyASites.gr %Q% !(name %in% names(pc_pas.gr))

nc_pas.gr <- (npc_pas.gr + 25) %&&%
    gr.end(nc_txs.gr, ignore.strand=FALSE, width=25)

other_pas.gr <- npc_pas.gr %Q% !(name %in% names(nc_pas.gr))
```

Annotation consists of `r length(pc_txs.gr)` protein coding and 
`r length(nc_txs.gr)` non-coding transcripts. Of the PolyASite cleavage sites, 
`r length(pc_pas.gr)` are near 3\'-ends of annotated protein coding transcripts,
`r length(nc_pas.gr)` are near 3\'-ends of annotated non-coding transcripts, and 
`r length(other_pas.gr)` are not near an annotated end.

## Plot Subsets
### Protein Coding
```{r fig.width=6, fig.height=4}
polyASites.df %>%
    filter(name %in% names(pc_pas.gr)) %>%
    ggplot(aes(x=as.character(score), fill=annot)) +
    geom_bar() +
    labs(x="Supporting Protocols",
         y="Cleavage Sites",
         fill="Annotation",
         title="GENCODE Protein Coding Sites") +
    theme_minimal()
```

### Non-Coding Transcripts
```{r fig.width=6, fig.height=4}
polyASites.df %>%
    filter(name %in% names(nc_pas.gr)) %>%
    ggplot(aes(x=as.character(score), fill=annot)) +
    geom_bar() +
    labs(x="Supporting Protocols",
         y="Cleavage Sites",
         fill="Annotation",
         title="GENCODE Non-Coding Sites") +
    theme_minimal()
```

### Not Annotated
```{r fig.width=6, fig.height=4}
polyASites.df %>%
    filter(name %in% names(other_pas.gr)) %>%
    ggplot(aes(x=as.character(score), fill=annot)) +
    geom_bar() +
    labs(x="Supporting Protocols",
         y="Cleavage Sites",
         fill="Annotation",
         title="Non-GENCODE Sites") +
    theme_minimal()
```

---

# Session Info
```{r sesh_info, echo=FALSE}
sessionInfo()
```
